### ISSUE TITLE
Mount failed with error: exit status 255, output: Unable to start blobfuse due to a lack of credentials

### ISSUE THREAD

**Issue by bartwitkowski** [kubernetes-sigs/blob-csi-driver#450]
Hi,
I use static creation of PV/PVC. Using Access Key mounting works but using SAS Tokens (with KV) it doesn't. Could you point me what I did wrong?
I've got the following error:

> I0608 08:32:07.484577       1 utils.go:114] GRPC call: /csi.v1.Node/NodeStageVolume
I0608 08:32:07.484612       1 utils.go:115] GRPC request: {"staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv-blob-test/globalmount","volume_capability":{"AccessType":{"Mount":{}},"access_mode":{"mode":5}},"volume_context":{"AzureStorageAuthType":"SAS","containerName":"test","keyVaultSecretName":"test-SasToken","keyVaultURL":"https://test-kv.vault.azure.net/","storageAccountName":"teststorage"},"volume_id":"uniqe-volumeid-test2-blob-pv"}
W0608 08:32:07.484817       1 blob.go:218] parsing volumeID(uniqe-volumeid-test2-blob-pv) return with error: error parsing volume id: "uniqe-volumeid-test2-blob-pv", should at least contain two #
I0608 08:32:07.484854       1 blob.go:263] volumeID(uniqe-volumeid-test2-blob-pv) authEnv: [AZURE_STORAGE_AUTH_TYPE=SAS]
I0608 08:32:07.743473       1 nodeserver.go:214] target /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv-blob-test/globalmount
protocol
volumeId uniqe-volumeid-test2-blob-pv
context map[AzureStorageAuthType:SAS containerName:test keyVaultSecretName:test-SasToken keyVaultURL:https://test-kv.vault.azure.net/ storageAccountName:teststorage]
mountflags []
mountOptions [--use-https=true]
args /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv-blob-test/globalmount --tmp-path=/mnt/uniqe-volumeid-test2-blob-pv#1623141127 --container-name=test --use-https=true
serverAddress teststorage.blob.core.windows.net
E0608 08:32:07.753109       1 nodeserver.go:224] Mount failed with error: exit status 255, output: Unable to start blobfuse due to a lack of credentials. Please check the readme for valid auth setups.
E0608 08:32:07.753278       1 utils.go:119] GRPC error: Mount failed with error: exit status 255, output: Unable to start blobfuse due to a lack of credentials. Please check the readme for valid auth setups.

Kubernetes 1.16.10
CSI blob v1.0.0

pv-csi.yaml
```apiVersion: v1
kind: PersistentVolume
metadata:
  name: pv-blob-test
spec:
  capacity:
    storage: 10Gi
  accessModes:
    - ReadWriteMany
  persistentVolumeReclaimPolicy: Retain  # "Delete" is not supported in static provisioning
  csi:
    driver: blob.csi.azure.com
    readOnly: false
    volumeHandle: uniqe-volumeid-test2-blob-pv
    volumeAttributes:
      containerName: test
      storageAccountName: teststorage
      keyVaultURL: https://test-kv.vault.azure.net/
      keyVaultSecretName: test-SasToken
      AzureStorageAuthType: SAS
```
pvc-csi.yaml
```kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: pvc-blob-test
spec:
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 10Gi
  volumeName: pv-blob-test
  storageClassName: ""
```
test app nginx-pod.yaml
```kind: Pod
apiVersion: v1
metadata:
  name: nginx-blob
spec:
  nodeSelector:
    "kubernetes.io/os": linux
  containers:
    - image: mcr.microsoft.com/oss/nginx/nginx:1.17.3-alpine
      name: nginx-blob
      volumeMounts:
        - name: blob01
          mountPath: "/mnt/blob"
  volumes:
    - name: blob01
      persistentVolumeClaim:
        claimName: pvc-blob-test
```

I've also tried this with CSI blob v1.1.0. There is no such error but the mount directory is empty and I've got problem with unmounting and error logs like "device or resource busy". Then, restarting csi-blob-node doesn't help and I had to restart whole Kubernetes VM.

**Comment by bartwitkowski** [comment]
At first I tried without this option - no success.

**Comment by andyzhangx** [comment]
if `AzureStorageAuthType: SAS` is specified, should create a secret like this:
```
kubectl create secret generic azure-secret --from-literal=azurestorageaccountname="xxx" --from-literal azurestorageaccountsastoken="xxx" --type=Opaque
```

The correct config format should be:
```yaml
    volumeAttributes:
      containerName: test
      storageAccountName: teststorage
      keyVaultURL: https://test-kv.vault.azure.net/
      keyVaultSecretName: test-SasToken
```

Pls provide node driver logs if that does not work.

**Comment by bartwitkowski** [comment]
It doesn't work. 
Using config:
```yaml
    volumeAttributes:
      containerName: test
      storageAccountName: teststorage
      keyVaultURL: https://test-kv.vault.azure.net/
      keyVaultSecretName: test-SasToken
```

Logs from csi-blob-node:

> I0608 10:13:26.907667       1 utils.go:115] GRPC request: {"staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv-blob-test/globalmount","volume_capability":{"AccessType":{"Mount":{}},"access_mode":{"mode":5}},"volume_context":{"containerName":"test","keyVaultSecretName":"test-SasToken","keyVaultURL":"https://test-kv.vault.azure.net/","storageAccountName":"teststorage"},"volume_id":"uniqe-volumeid-test2-blob-pv"}
W0608 10:13:26.908021       1 blob.go:218] parsing volumeID(uniqe-volumeid-test2-blob-pv) return with error: error parsing volume id: "uniqe-volumeid-test2-blob-pv", should at least contain two #
I0608 10:13:26.908042       1 blob.go:263] volumeID(uniqe-volumeid-test2-blob-pv) authEnv: []
I0608 10:13:27.404208       1 nodeserver.go:214] target /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv-blob-test/globalmount
protocol
volumeId uniqe-volumeid-test2-blob-pv
context map[containerName:test keyVaultSecretName:test-SasToken keyVaultURL:https://test-kv.vault.azure.net/ storageAccountName:teststorage]
mountflags []
mountOptions [--use-https=true]
args /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv-blob-test/globalmount --tmp-path=/mnt/uniqe-volumeid-test2-blob-pv#1623147207 --container-name=test --use-https=true
serverAddress teststorage.blob.core.windows.net
E0608 10:13:27.505227       1 nodeserver.go:224] Mount failed with error: exit status 255, output: Unable to start blobfuse due to a lack of credentials. Please check the readme for valid auth setups.

**Comment by andyzhangx** [comment]
have you stored the sas token as secret in `test-SasToken`?

**Comment by bartwitkowski** [comment]
Yes, the sas token is in required KV as a secred.
With CSI blob v 1.1.0 I don't have such "lack of credentials" problem. But I have some others, that may be connected to K8s version - 1.16.10 which is not supported in 1.1.0 as I understand.

**Comment by andyzhangx** [comment]
> Yes, the sas token is in required KV as a secred.
> With CSI blob v 1.1.0 I don't have such "lack of credentials" problem. But I have some others, that may be connected to K8s version - 1.16.10 which is not supported in 1.1.0 as I understand.

what's the error msg with v1.1.0?

**Comment by bartwitkowski** [comment]
Well, the mounting process looks ok, logs are clean, BUT the /mnt/blob directory in the pod is empty. So actually the mounting failed. And when I want to remove the pod, in the csi-blob-node there is log: "device or resource busy"
Full log:

> I0609 13:20:08.765220       1 utils.go:115] GRPC request: {"staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv-blob-test/globalmount","volume_capability":{"AccessType":{"Mount":{}},"access_mode":{"mode":5}},"volume_context":{"containerName":"test","keyVaultSecretName":"test-SasToken","keyVaultURL":"https://test-kv.vault.azure.net/","storageAccountName":"teststorage"},"volume_id":"uniqe-volumeid-test2-blob-pv"}
W0609 13:20:08.765488       1 blob.go:227] parsing volumeID(uniqe-volumeid-test2-blob-pv) return with error: error parsing volume id: "uniqe-volumeid-test2-blob-pv", should at least contain two #
I0609 13:20:08.765513       1 blob.go:272] volumeID(uniqe-volumeid-test2-blob-pv) authEnv: []
I0609 13:20:09.225886       1 nodeserver.go:267] target /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv-blob-test/globalmount
protocol
volumeId uniqe-volumeid-test2-blob-pv
context map[containerName:test keyVaultSecretName:test-SasToken keyVaultURL:https://test-kv.vault.azure.net/ storageAccountName:teststorage]
mountflags []
mountOptions [--use-https=true]
args /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv-blob-test/globalmount --tmp-path=/mnt/uniqe-volumeid-test2-blob-pv#1623244809 --container-name=test --use-https=true
serverAddress teststorage.blob.core.windows.net
I0609 13:20:09.225922       1 nodeserver.go:140] mounting blobfuse inside driver
I0609 13:20:09.715227       1 nodeserver.go:306] volume(uniqe-volumeid-test2-blob-pv) mount on "/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv-blob-test/globalmount" succeeded
I0609 13:20:09.715268       1 utils.go:121] GRPC response: {}
I0609 13:20:09.725484       1 utils.go:114] GRPC call: /csi.v1.Node/NodePublishVolume
I0609 13:20:09.725503       1 utils.go:115] GRPC request: {"staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv-blob-test/globalmount","target_path":"/var/lib/kubelet/pods/d427ea5e-f137-461c-bfc9-17e576dff635/volumes/kubernetes.io~csi/pv-blob-test/mount","volume_capability":{"AccessType":{"Mount":{}},"access_mode":{"mode":5}},"volume_context":{"containerName":"test","csi.storage.k8s.io/ephemeral":"false","csi.storage.k8s.io/pod.name":"nginx-blob","csi.storage.k8s.io/pod.namespace":"mlclassifier","csi.storage.k8s.io/pod.uid":"d427ea5e-f137-461c-bfc9-17e576dff635","csi.storage.k8s.io/serviceAccount.name":"default","keyVaultSecretName":"test-SasToken","keyVaultURL":"https://test-kv.vault.azure.net/","storageAccountName":"teststorage"},"volume_id":"uniqe-volumeid-test2-blob-pv"}
I0609 13:20:09.725776       1 nodeserver.go:94] NodePublishVolume: volume uniqe-volumeid-test2-blob-pv mounting /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv-blob-test/globalmount at /var/lib/kubelet/pods/d427ea5e-f137-461c-bfc9-17e576dff635/volumes/kubernetes.io~csi/pv-blob-test/mount with mountOptions: [bind]
I0609 13:20:09.725840       1 mount_linux.go:175] Mounting cmd (mount) with arguments ( -o bind /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv-blob-test/globalmount /var/lib/kubelet/pods/d427ea5e-f137-461c-bfc9-17e576dff635/volumes/kubernetes.io~csi/pv-blob-test/mount)
I0609 13:20:09.772492       1 mount_linux.go:175] Mounting cmd (mount) with arguments ( -o bind,remount /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv-blob-test/globalmount /var/lib/kubelet/pods/d427ea5e-f137-461c-bfc9-17e576dff635/volumes/kubernetes.io~csi/pv-blob-test/mount)
I0609 13:20:09.774069       1 nodeserver.go:110] NodePublishVolume: volume uniqe-volumeid-test2-blob-pv mount /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv-blob-test/globalmount at /var/lib/kubelet/pods/d427ea5e-f137-461c-bfc9-17e576dff635/volumes/kubernetes.io~csi/pv-blob-test/mount successfully
I0609 13:20:09.774093       1 utils.go:121] GRPC response: {}
I0609 13:25:10.370128       1 utils.go:114] GRPC call: /csi.v1.Node/NodeUnpublishVolume
I0609 13:25:10.370148       1 utils.go:115] GRPC request: {"target_path":"/var/lib/kubelet/pods/d427ea5e-f137-461c-bfc9-17e576dff635/volumes/kubernetes.io~csi/pv-blob-test/mount","volume_id":"uniqe-volumeid-test2-blob-pv"}
I0609 13:25:10.370245       1 nodeserver.go:163] NodeUnpublishVolume: unmounting volume uniqe-volumeid-test2-blob-pv on /var/lib/kubelet/pods/d427ea5e-f137-461c-bfc9-17e576dff635/volumes/kubernetes.io~csi/pv-blob-test/mount
W0609 13:25:10.370276       1 mount_helper_common.go:129] Warning: "/var/lib/kubelet/pods/d427ea5e-f137-461c-bfc9-17e576dff635/volumes/kubernetes.io~csi/pv-blob-test/mount" is not a mountpoint, deleting
E0609 13:25:10.370325       1 utils.go:119] GRPC error: rpc error: code = Internal desc = failed to unmount target "/var/lib/kubelet/pods/d427ea5e-f137-461c-bfc9-17e576dff635/volumes/kubernetes.io~csi/pv-blob-test/mount": remove /var/lib/kubelet/pods/d427ea5e-f137-461c-bfc9-17e576dff635/volumes/kubernetes.io~csi/pv-blob-test/mount: device or resource busy

And the NodeUnpublishVolume retries all over...

**Comment by andyzhangx** [comment]
could you ssh to that node to check whether blobfuse mount succeeded:
```
ls /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv-blob-test/globalmount
mount | grep blobfuse | uniq
```

Not sure why sas token does not work. What if you put a storage account key in key vault secret, does it work?

**Comment by bartwitkowski** [comment]
Commands from csi-blob-node:

> ls /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv-blob-test/globalmount
testa

And yes, file "testa" was created inside the nginx-blob container.
Command "mount | grep blobfuse | uniq" returns nothing - there is no blobfuse mount type.

Also, when I create a new nginx-blob container instance now I don't see the "testa" file inside.

Storage account key has worked previously.

**Comment by bartwitkowski** [comment]
Further troubleshooting:
CSI blob v1.0.0 and v1.1.0 + storage access key works and i see blobfuse mount in the container:
> blobfuse on /mnt/blob type fuse (rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other,max_read=131072)

CSI blob v.1.0.0 +  SAS token in KV log:
> nodeserver.go:224] Mount failed with error: exit status 255, output: Unable to start blobfuse due to a lack of credentials. Please check the readme for valid auth setups.

CSI blob v.1.1.0 +  SAS token in KV mounting looks ok but instead of blobfuse mount I see ext4 and the /mnt/blob direcotry is empty:
> /dev/sda1 on /mnt/blob type ext4 (rw,relatime,discard,data=ordered)

And I cannot gently delete the nginx-blob pod, csi-blob-node log:
> device or resource busy

All in the same AKS v 1.16.10

**Comment by bartwitkowski** [comment]
Hey,
Someone - any clues?
What should I check more?

**Comment by andyzhangx** [comment]
is it possible to try store sas token in k8s secret, follow guide here:

```
kubectl create secret generic azure-secret --from-literal=azurestorageaccountname="xxx" --from-literal azurestorageaccountsastoken="xxx" --type=Opaque
```

https://github.com/kubernetes-sigs/blob-csi-driver/blob/master/deploy/example/e2e_usage.md#option2-use-secret

there is a bug: [read keyvault secret only supports service principal(managed identity does not work) ](https://github.com/kubernetes-sigs/blob-csi-driver/issues/453), not sure whether it's related, what's your cluster identity type, service principal or managed identity?

**Comment by bartwitkowski** [comment]
OK, I'll try with SAS token in a secret and will get back to you,
And yes, we are using MSI.
I saw that bug, but even if it is the case, then the fix will be in the newest version of your driver which won't be compatible with AKS 1.16, right?

**Comment by andyzhangx** [comment]
> OK, I'll try with SAS token in a secret and will get back to you,
> And yes, we are using MSI.
> I saw that bug, but even if it is the case, then the fix will be in the newest version of your driver which won't be compatible with AKS 1.16, right?

correct

**Comment by andyzhangx** [comment]
I have successfully run a pod with blobfuse mount reading from SAS token in keyvault secret, I think the original issue is due to this bug(https://github.com/kubernetes-sigs/blob-csi-driver/issues/453), it's only fixed in master branch(require k8s 1.18 version). One tricky thing I hit is that if SAS token is expired(or invalid, in my case it's expired), blobfuse mount would also return success, and finally it was mounted to a local disk.

```
apiVersion: v1
kind: PersistentVolume
metadata:
  name: pv-blob-test4
spec:
  capacity:
    storage: 10Gi
  accessModes:
    - ReadWriteMany
  persistentVolumeReclaimPolicy: Retain  # "Delete" is not supported in static provisioning
  csi:
    driver: blob.csi.azure.com
    readOnly: false
    volumeHandle: uniqe-volumeid-test4-blob-pv
    volumeAttributes:
      containerName: test
      storageAccountName: xxx
      keyVaultURL: https://xxx.vault.azure.net/
      keyVaultSecretName: test-SasToken2
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: pvc-blob-test4
spec:
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 10Gi
  volumeName: pv-blob-test4
  storageClassName: ""
---
apiVersion: v1
kind: Pod
metadata:
  name: nginx-blob6
spec:
  nodeSelector:
    "kubernetes.io/os": linux
  containers:
    - image: mcr.microsoft.com/oss/nginx/nginx:1.17.3-alpine
      name: nginx-blob
      volumeMounts:
        - name: blob01
          mountPath: "/mnt/blob"
  volumes:
    - name: blob01
      persistentVolumeClaim:
        claimName: pvc-blob-test4
```
